### PR TITLE
rejected edits trailing whitespaces 

### PR DESCRIPTION
**Pull request by panyamkeerthana**
## Description

Earlier trailing whitespace was stripped in streamDiffLines.ts before generating diffs, causing mismatches when rejecting a diff in the editor.
This fix updates linesMatchPerfectly to compare lines using trimEnd(), ensuring that trailing whitespace differences do not interfere with accurate diff rejection.

## Screenshots

Screenshot 2025-03-14 at 3 14 53 PM

Screenshot 2025-03-14 at 3 14 28 PM

https://github.com/Granite-Code/granite-code/issues/50

### CODE REVIEW

File: extensions/vscode/src/quickEdit/EditDecorationManager.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@ class EditDecorationManager {
   constructor(context: vscode.ExtensionContext) {
     this.decorationType = vscode.window.createTextEditorDecorationType({
       backgroundColor: new vscode.ThemeColor(
-        "editor.inactiveSelectionBackground",
+        "editor.selectionHighlightBackground",
```
Comment (owtaylor) [inline extensions/vscode/src/quickEdit/EditDecorationManager.ts]: A stow-away

File: core/diff/util.ts [hunk 2]
```diff
@@ -16,7 +16,11 @@ export type MatchLineResult = {
 };
 
 function linesMatchPerfectly(lineA: string, lineB: string): boolean {
-  return lineA === lineB && lineA !== "";
+
+//checks if two lines match exactly ignoring trailing whitespace differences
+//this prevents false mismatches in diff handling when trailing spaces are stripped
+
+  return lineA.trimEnd() === lineB.trimEnd() && lineA !== "";
```
Comment (owtaylor) [inline core/diff/util.ts]: This fixes up one case where we are comparing oldLines to newLines, but checking over the code, this isn't the only place where we might have issues:

 - in `matchLine()`, after calling `linesMatchPerfectly()`, we make more references to `oldLines[i]` that potentially might need changes.
 - In streamDiff(), we have:
```
         if (oldLinesCopy[0] !== newLine) {
          yield { type: "new", line: newLine };
        } else {
          isLineRemoval = true;
        }
```
   which looks suspicious, but I'm pretty sure can never actually be hit: if `oldLinesCopy[0] === newLine` then `isPerfectMatch` should have been true and the case will be `"same"` not `"old"`. Needs double checking.

But the takeaway for me here is that the test cases need updating to test different cases where lines differ only in trailing whitespace. From the `continue/core` directory, you can run:

```
$ npm test  diff/streamDiff.test.ts
```

to run streamDiff test cases. Add a test case to check the `linesMatchPerfectly` case, add a test case case that exercises:       `// This is a way to fix indentation, but only for sufficiently long lines to avoid matching whitespace or short lines` when the indentation changes *and* there is a change in whitespace.

Comment (panyamkeerthana) [reply core/diff/util.ts]: Thanks for the review, I have made the following changes:

-Added a test case for oldLinesCopy[0] === newLine that confirmed it’s already handled as "same" or "old", so the else statement was unnecessary and removed.
-Trimmed oldLines throughout and added a test to ensure trailing whitespaces are correctly treated as "same", preventing incorrect diffs.
-Added a test to verify that indentation changes in long lines don’t cause unnecessary diffs.

File: core/diff/streamDiff.test.ts [hunk 3]
```diff
@@ -223,6 +223,51 @@ describe("streamDiff(", () => {
...
+
+    expect(streamDiffs).toEqual([
+      { type: "same", line: "first item  " },
+      { type: "same", line: "second arg " },
+      { type: "same", line: "third param" },
+    ]);
+  });
+
+  test("ignores indentation changes for sufficiently long lines", async () => {
+    const oldLines = ["   first item", " second arg", "third param", " fourth val "];
```
Comment (owtaylor) [inline core/diff/streamDiff.test.ts]: I would try to test this more comprehensively - to test that indentation changes on short lines are considered changes, until one line with length > 8 is encountered, then they are ignored, even for short lines.

Importantly, I'd also check here that the same behavior is seen when the old and new lines differ in trailing whitespace - that's the point of adding the check as part of this patch, after all. 

**Something** like:

```
 test.each([false, true])(
   `ignores identation changes for sufficiently long lines",
   async (trailingWhitespace) => {
      [...]
      if (trailingWhitespace)
         oldLines = oldLines.map(line => line + " ");
      [...]
   });
```

I'm sure there's a more stylish way to do that.

File: core/diff/util.ts [hunk 4]
```diff
@@ -62,24 +62,27 @@ export function matchLine(
...
+    //this ensures trailing spaces don't affect matching.
+    const oldLineTrimmed = oldLines[i].trimEnd();
     // Don't match end bracket lines if too far away
     if (i > 4 && isEndBracket) {
       return { matchIndex: -1, isPerfectMatch: false, newLine };
     }
 
-    if (linesMatchPerfectly(newLine, oldLines[i])) {
+    if (linesMatchPerfectly(newLine, oldLineTrimmed)) {
       return { matchIndex: i, isPerfectMatch: true, newLine };
```
Comment (owtaylor) [inline core/diff/util.ts:76]: I think this probably should be oldLines[i] not newLine? The goal here should be that when we return `same` - we actually have the old line, not a version that differs in trailing whitespace. First try to make a test case that fails, then change this.

File: core/diff/util.ts [hunk 5]
```diff
@@ -62,24 +62,27 @@ export function matchLine(
...
       if (
-        newLine.trimStart() === oldLines[i].trimStart() &&
+        newLine.trimStart() === oldLineTrimmed.trimStart() &&
         (permissiveAboutIndentation || newLine.trim().length > 8)
       ) {
         return {
           matchIndex: i,
           isPerfectMatch: true,
-          newLine: oldLines[i],
+          newLine: oldLineTrimmed,
```
Comment (owtaylor) [inline core/diff/util.ts]: Similarly, I think this should be oldLines[i]?

File: core/diff/streamDiff.ts [hunk 6]
```diff
@@ -63,8 +63,6 @@ export async function* streamDiff(
 
         if (oldLinesCopy[0] !== newLine) {
           yield { type: "new", line: newLine };
-        } else {
-          isLineRemoval = true;
         }
```
Comment (owtaylor) [inline core/diff/streamDiff.ts]: You should remove the if/else entirely and just leave the yield, then since isLineRemoval can never be set, it needs to be removed as well.

File: core/diff/streamDiff.test.ts [hunk 7]
```diff
@@ -223,6 +223,51 @@ describe("streamDiff(", () => {
     await expectDiff("fastapi-tabs-vs-spaces.py");
   });
 
+
+  test(" condition where oldLinesCopy[0]=== newLine", async () => {
```
Comment (owtaylor) [inline core/diff/streamDiff.test.ts]: I think this was useful during development of your patch, but since this code ends up removed in your patches, I'd suggest removing the test case as well.

File: core/diff/util.ts (hunk 4 above)
Comment (panyamkeerthana) [reply core/diff/util.ts:76]: The test cases I ran for this scenario passed. In streamDiff():

```
case "same":
    yield { type, line: oldLinesCopy.shift()! };
    break;
```
So, regardless of what matchLine() returns for newLine, the line that is actually output is the one removed from oldLinesCopy (which is the original old line with the trailing whitespaces). But I can change it to oldLine[i] to maintain logic and uniformity

File: core/diff/util.ts (hunk 5 above)
Comment (panyamkeerthana) [reply core/diff/util.ts]: I'll change that to oldLine[i] here as well

File: core/edit/streamDiffLines.ts [hunk 8]
```diff
@@ -79,7 +79,7 @@ export async function* streamDiffLines(
   }
 
   // Trim end of oldLines, otherwise we have trailing \r on every line for CRLF files
-  oldLines = oldLines.map((line) => line.trimEnd());
+  oldLines = oldLines.map((line) => line);
```
Comment (owtaylor) [inline core/edit/streamDiffLines.ts]: This can just go away, along with the comment

File: core/diff/streamDiff.ts [hunk 9]
```diff
@@ -35,17 +35,14 @@ export async function* streamDiff(
 
     let type: DiffLineType;
 
-    let isLineRemoval = false;
     const isNewLine = matchIndex === -1;
 
     if (isNewLine) {
       type = "new";
     } else {
-      // Insert all deleted lines before match
```
Comment (owtaylor) [inline core/diff/streamDiff.ts]: Should this comment be removed?

File: core/diff/streamDiff.test.ts [hunk 10]
```diff
@@ -223,6 +223,68 @@ describe("streamDiff(", () => {
...
+        "a long enough line",
+        "  indented short",
+        "final line",
+      ];
+
+      let newLines = [
+        "short",
+        "middle",
+        "a long enough line",
+        "indented short",
```
Comment (owtaylor) [inline core/diff/streamDiff.test.ts]: "indented short".length is going to be > 8 - this needs to be "short2" or something to actually test that handlng.

File: core/diff/util.ts [hunk 11]
```diff
@@ -62,18 +62,21 @@ export function matchLine(
...
-      return { matchIndex: i, isPerfectMatch: true, newLine };
+    if (linesMatchPerfectly(newLine, oldLineTrimmed)) {
+      return { matchIndex: i, isPerfectMatch: true, newLine: oldLines[i] };
     }
-    if (linesMatch(newLine, oldLines[i], i)) {
+    if (linesMatch(newLine, oldLineTrimmed, i)) {
       // This is a way to fix indentation, but only for sufficiently long lines to avoid matching whitespace or short lines
       if (
-        newLine.trimStart() === oldLines[i].trimStart() &&
+        newLine.trimStart() === oldLineTrimmed.trimStart() &&
```
Comment (owtaylor) [inline core/diff/util.ts]: I don't think we can *assume* that newLine is trimmed at the end, so this should be newLine.trim()? [make sure the test case tests this - one line that is supposed to not reindented should have a trailing space on the new]

**Review by owtaylor — COMMENTED** [review]
Just a few more comments

File: core/diff/util.ts (hunk 11 above)
Comment (panyamkeerthana) [reply core/diff/util.ts]: Instead, as discussed I created newLineTrimmed.

` newLineTrimmed.trimStart() === oldLineTrimmed.trimStart()`

File: core/diff/streamDiff.ts (hunk 9 above)
Comment (panyamkeerthana) [reply core/diff/streamDiff.ts]: No I might have accidentally removed it, my bad

File: core/diff/util.ts (hunk 4 above)
Comment (owtaylor) [reply core/diff/util.ts:76]: Turned out that this very much needed to be newLine - the calling code relies on this only being different from the passed in newline when the leading indentation changed. Sorry for the confusion!

File: core/diff/streamDiff.ts (hunk 9 above)
Comment (owtaylor) [reply core/diff/streamDiff.ts]: If you add this back, the rest looks great to me!

File: core/diff/util.ts (hunk 4 above)
Comment (panyamkeerthana) [reply core/diff/util.ts:76]: All this testing has only made the whitespace and indentation handling more robust!

File: core/diff/streamDiff.ts (hunk 9 above)
Comment (panyamkeerthana) [reply core/diff/streamDiff.ts]: Done!

**Comment by owtaylor** [comment]
Looks good to me now!

**Review by tomasz-stefaniak — APPROVED** [review]
Thank you for contributing @panyamkeerthana, and @owtaylor for the thorough review!
